app/src/main/java/ch/epfl/sdp/storage/ConcreteManager.java

```diff
@@ -0,0 +1,193 @@
+package ch.epfl.sdp.storage;
+
+import android.content.Context;
+import android.os.AsyncTask;
+
+import java.io.BufferedReader;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.util.Collections;
+import java.util.Map;
+import java.util.TreeMap;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+
+/**
+ * Implements a StorageManager with cache (asynchronously preloaded)
+ * @param <A>
+ * @param <B>
+ */
+public class ConcreteManager<A extends Comparable<A>, B> implements StorageManager<A, B> {
+
+    private boolean isDeleted = false;
+    private File file;
+    private FileWriter writer = null;
+
+    private volatile Map<A, B> cache;
+
+    private Function<String, A> stringToA;
+    private Function<String, B> stringToB;
+
+    private AtomicBoolean loadingCache;
+    private AtomicBoolean cacheOk;
+
+    private static final String SEPARATOR = ",";
+    private static final String LINE_END = "\n";
+
+    public ConcreteManager(Context context, String filename, Function<String, A> convertToA, Function<String, B> convertToB) {
+        if (convertToA == null || convertToB == null) {
+            throw new IllegalArgumentException();
+        }
+
+        stringToA = convertToA;
+        stringToB = convertToB;
+
+        file = new File(context.getFilesDir(), filename);
+        if (file.isDirectory()) {
+            throw new IllegalArgumentException("Need a file and not a directory");
+        }
+
+        if (!file.exists()) {
+            try {
+                file.createNewFile();
+            } catch (IOException e) {
+                throw new IllegalArgumentException("Unable to create a new file, must select existing one");
+            }
+        }
+        if (!file.canWrite() || !file.canRead()) {
+            throw new IllegalArgumentException("Cannot read&write on the specified file");
+        }
+
+        cache = new TreeMap<>();
+
+        loadingCache = new AtomicBoolean(false);
+        cacheOk = new AtomicBoolean(false);
+
+        AsyncTask.execute(() -> {
+                boolean result = loadCache();
+                cacheOk.set(result);
+                loadingCache.set(true);
+        });
+    }
+
+    private void checkCacheStatus() {
+        if (loadingCache.get() && !cacheOk.get()) {
+            throw new IllegalStateException("Could not perform initial cache loading");
+        }
+
+        // Block to wait cache loading
+        while (!loadingCache.get()) { }
```

Comment:
No way to avoid a busy wait? Maybe return CompletableFuture\<Void\>

Reply:
I deemed it was not necessary since the cache loading is started when the object is created and the cache is supposed to be small. Furthermore, the write and read methods are not supposed to be called from UI thread